### ISSUE TITLE
[Question] Clarification on asymmetric clipping threshold in GRPO loss

### ISSUE THREAD

**Issue by SriGoli-0123** [natolambert/rlhf-book#427]
Hi @natolambert,

Thank you for putting this repository and book together! 

I have a question regarding the clipping threshold values used in the GRPO loss function implementation. In the code snippet below:

```python
def forward(self, log_probs: torch.Tensor, experience: Experience, **kwargs) -> torch.Tensor:
        # Policy loss with clipping
        ratio = (log_probs - experience.log_probs_old).exp()
        unclipped_term = ratio * experience.advantages
        clipped_term = (
            ratio.clamp(1 - self.clip_eps_lo, 1 + self.clip_eps_hi) * experience.advantages
        )
        policy_loss = -torch.min(unclipped_term, clipped_term)

        # Optional KL penalty
        if self.beta:
            kl_loss = get_approx_kl(
                self.kl_estimator, log_probs, experience.log_probs_ref, experience.action_mask
            )
        else:
            kl_loss = torch.tensor(0.0, device=log_probs.device, dtype=torch.float32)

        loss = policy_loss + self.beta * kl_loss
        loss = masked_mean(loss, mask=experience.action_mask, dim=-1).mean(dim=0)
        return loss
```

The implementation uses two distinct clipping thresholds, clip_eps_lo and clip_eps_hi. However, the original DeepSeek GRPO paper specifies a single ϵ value for symmetric clipping (similar to standard PPO), whereas asymmetric thresholds are explicitly featured in papers like CISPO.

Could you clarify if using asymmetric clipping in the other Loss functions as well, was a deliberate optimization choice for this codebase, or if it's meant to support a generalized PPO-style clipping interface?

Thanks for your time!

Best regards,
Harshith

**Comment by natolambert** [comment]
Thanks for flagging this. Yes, the separate `clip_eps_lo` / `clip_eps_hi` fields are there to support DAPO / clip-higher-style variants, where the lower and upper clipping thresholds can differ. See the DAPO paper: https://arxiv.org/abs/2503.14476

For the default GRPO config, those values are equal (`0.2` and `0.2`), so the implementation reduces to the standard symmetric GRPO clipping range from the DeepSeek GRPO objective. Closing since this is expected behavior rather than a bug.
